src/session/content/background.rs

```diff
@@ -0,0 +1,397 @@
+    float dp3 = distance(uv, p3);
+    float dp4 = distance(uv, p4);
+    float minD = min(dp1, min(dp2, min(dp3, dp4)));
+    float p = 5.0;
+    dp1 = pow(1.0 - (dp1 - minD), p);
+    dp2 = pow(1.0 - (dp2 - minD), p);
+    dp3 = pow(1.0 - (dp3 - minD), p);
+    dp4 = pow(1.0 - (dp4 - minD), p);
+    float sumDp = dp1 + dp2 + dp3 + dp4;
+
+    vec3 color = (color1 * dp1 + color2 * dp2 + color3 * dp3 + color4 * dp4) / sumDp;
+    fragColor = vec4(color, 1.0);
+}
+"#
+.as_bytes();
+
+const PATTERN_SHADER: &[u8] = r#"
+precision highp float;
+precision highp sampler2D;
+
+uniform bool dark;
+uniform sampler2D u_texture1;
+uniform sampler2D u_texture2;
+
+bool approxEq(float a, float b) {
+    return abs(a - b) < 0.05;
+}
+
+void mainImage(out vec4 fragColor,
+    in vec2 fragCoord,
+    in vec2 resolution,
+    in vec2 uv) {
+
+    vec4 messages = GskTexture(u_texture1, uv);
+    vec4 pattern = GskTexture(u_texture2, uv);
+
+    float message_alpha = dark ? 0.9 : 0.8;
+    float event_alpha = dark ? 0.8 : 0.2;
+
+    // We don't need to draw pattern under semi-transparent messages
+    // But we need to draw it under antialized corners
+    if (
+        approxEq(messages.a, message_alpha) ||
+        approxEq(messages.a, event_alpha) ||
+        messages.a == 1.0
+    ) {
+        fragColor = messages;
+    } else {
+        vec4 pattern_color;
+
+        if (dark) {
+            float alpha = 1.0 - pattern.a * 0.3;
+            pattern_color = vec4(vec3(30.0 / 255.0), 1.0) * alpha;
+        } else {
+            float alpha = pattern.a * 0.1;
+            pattern_color = vec4(vec3(0.0), 1.0) * alpha;
+        }
+
+        // blend colors with premultiplied alpha
+        fragColor = messages + pattern_color * (1.0 - messages.a);
+    }
+}
+"#
+.as_bytes();
+
+mod imp {
+    use super::*;
+    use gtk::glib::once_cell::sync::{Lazy, OnceCell};
+    use gtk::glib::subclass::Signal;
+    use gtk::glib::{self, clone};
+
+    #[derive(PartialEq, Clone, Copy)]
+    struct GradientCacheParams {
+        size: (f32, f32),
+        colors: [graphene::Vec3; 4],
+        phase: u32,
+    }
+
+    #[derive(Default)]
+    pub struct Background {
+        pub(super) gradient_cache: RefCell<Option<gdk::Texture>>,
+        pub(super) last_size: Cell<(f32, f32)>,
+
+        pub(super) shaders: RefCell<Option<Option<[gsk::GLShader; 2]>>>,
+        pub(super) pattern: OnceCell<gdk::Texture>,
+
+        pub(super) progress: Cell<f32>,
+        pub(super) phase: Cell<u32>,
+
+        pub(super) dark: Cell<bool>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for Background {
+        const NAME: &'static str = "ComponentsBackground";
+        type Type = super::Background;
+        type ParentType = adw::Bin;
+    }
+
+    impl ObjectImpl for Background {
+        fn constructed(&self) {
+            self.parent_constructed();
+
+            let obj = self.obj();
+
+            obj.set_vexpand(true);
+            obj.set_hexpand(true);
```

Comment:
This should be set by users of this widget (`ChatHistory` in this case), not here.